File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotThrowFromDestructors.vb
Imports System

Namespace Tests.Diagnostics
    Class Program

        Public Sub Foo()
            Throw New Exception()
        End Sub

        Public Sub FooNoParens
            Throw New Exception()
        End Sub

        Public Sub New()
            Throw New Exception()
        End Sub

        Protected Overrides Overloads Sub fiNAliZE()
            Throw New Exception() ' Noncompliant {{Remove this 'Throw' statement.}}
'           ^^^^^^^^^^^^^^^^^^^^^
            Try
                Throw New Exception() ' Noncompliant
            Catch __unusedException1__ As Exception
            End Try

            Try
                Foo()
            Catch __unusedException1__ As Exception
                Throw ' Noncompliant
            End Try
        End Sub
    end class

    Class NoViolation1
        Protected Sub Finalize(i As Integer)
            Throw New Exception()
        End Sub
    End Class
    Class NoViolation2
        Sub Finalize()
            Throw New Exception()
        End Sub
    End Class
    Class NoViolation3
        Protected Sub Finalize(Of T)()
            Throw New Exception()
        End Sub
    End Class
    Class NoViolation4
        Protected Function Finalize() As Integer
            Throw New Exception()
        End Function
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/DoNotThrowFromDestructors.cs
using System;

namespace Tests.Diagnostics
{
    class Program
    {
        public void Foo()
        {
            throw new Exception(); // Compliant
        }

        public Program()
        {
            throw new Exception(); // Compliant, ctor covered by S3877
        }

        ~Program()
        {
            throw new Exception(); // Noncompliant {{Remove this 'throw' statement.}}
//          ^^^^^^^^^^^^^^^^^^^^^^

            void Inner()
            {
                throw new Exception(); // Noncompliant, C# 7+
            };

            try
            {
                throw new Exception(); // Noncompliant, generally a bad idea to throw and catch in the same method
            }
            catch (Exception)
            {
            }

            try
            {
                Foo();
            }
            catch (Exception)
            {
                throw; // Noncompliant, rethrowing has the same effect as throwing
            }
        }
    }
}
